// zksync/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.operations = exports.crypto = exports.utils = exports.types = exports.wallet = exports.EthMessageSigner = exports.closestPackableTransactionFee = exports.closestPackableTransactionAmount = exports.No2FAWalletSigner = exports.Create2WalletSigner = exports.Signer = exports.SyncProvider = exports.getDefaultRestProvider = exports.RestProvider = exports.getDefaultProvider = exports.ETHProxy = exports.Provider = exports.RemoteWallet = exports.submitSignedTransactionsBatch = exports.submitSignedTransaction = exports.ETHOperation = exports.Transaction = exports.Wallet = void 0;
var wallet_1 = require("./wallet");
Object.defineProperty(exports, "Wallet", { enumerable: true, get: function () { return wallet_1.Wallet; } });
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return wallet_1.Transaction; } });
Object.defineProperty(exports, "ETHOperation", { enumerable: true, get: function () { return wallet_1.ETHOperation; } });
Object.defineProperty(exports, "submitSignedTransaction", { enumerable: true, get: function () { return wallet_1.submitSignedTransaction; } });
Object.defineProperty(exports, "submitSignedTransactionsBatch", { enumerable: true, get: function () { return wallet_1.submitSignedTransactionsBatch; } });
var remote_wallet_1 = require("./remote-wallet");
Object.defineProperty(exports, "RemoteWallet", { enumerable: true, get: function () { return remote_wallet_1.RemoteWallet; } });
var provider_1 = require("./provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return provider_1.Provider; } });
Object.defineProperty(exports, "ETHProxy", { enumerable: true, get: function () { return provider_1.ETHProxy; } });
Object.defineProperty(exports, "getDefaultProvider", { enumerable: true, get: function () { return provider_1.getDefaultProvider; } });
var rest_provider_1 = require("./rest-provider");
Object.defineProperty(exports, "RestProvider", { enumerable: true, get: function () { return rest_provider_1.RestProvider; } });
Object.defineProperty(exports, "getDefaultRestProvider", { enumerable: true, get: function () { return rest_provider_1.getDefaultRestProvider; } });
var provider_interface_1 = require("./provider-interface");
Object.defineProperty(exports, "SyncProvider", { enumerable: true, get: function () { return provider_interface_1.SyncProvider; } });
var signer_1 = require("./signer");
Object.defineProperty(exports, "Signer", { enumerable: true, get: function () { return signer_1.Signer; } });
Object.defineProperty(exports, "Create2WalletSigner", { enumerable: true, get: function () { return signer_1.Create2WalletSigner; } });
Object.defineProperty(exports, "No2FAWalletSigner", { enumerable: true, get: function () { return signer_1.No2FAWalletSigner; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "closestPackableTransactionAmount", { enumerable: true, get: function () { return utils_1.closestPackableTransactionAmount; } });
Object.defineProperty(exports, "closestPackableTransactionFee", { enumerable: true, get: function () { return utils_1.closestPackableTransactionFee; } });
var eth_message_signer_1 = require("./eth-message-signer");
Object.defineProperty(exports, "EthMessageSigner", { enumerable: true, get: function () { return eth_message_signer_1.EthMessageSigner; } });
exports.wallet = __importStar(require("./wallet"));
exports.types = __importStar(require("./types"));
exports.utils = __importStar(require("./utils"));
exports.crypto = __importStar(require("./crypto"));
exports.operations = __importStar(require("./operations"));
require("./withdraw-helpers");
